enhancements/network/ovn-multi-external-gateway.md

```diff
@@ -0,0 +1,178 @@
+
+### Workflow Description
+The workflow follows the behavior of a namespaced CR.
+
+
+### API Extensions
+
+A new namespaced-scoped CRD is created to represent the multiple external gateways configuration. The `spec` structure exposes the following fields:
+* `staticGateways` contains a slice of IPv4 or IPv6 addresses that are configured as static external gateways. It maps to the `k8s.ovn.org/routing-external-gws` annotation.
+* `routingNamespaces` contains a slice of the namespaces linked to this gateway configuration. It maps to the `k8s.ovn.org/routing-namespaces` annotation.
+* `podGateways` allows defining the selector and configuration criteria for the gateway pods. This configuration contains the following fields:
+  * `network` references the multus network configuration that matches the same name in the pod that will be used for external traffic. If the field is empty or undefined, the pod's IP will be used instead. This matches the current behavior used with annotations. This field maps to the annotation `k8s.ovn.org/routing-network`.
+  * `bfdEnabled` determines if Bi-Directional Forwarding Detection (BFD) is supported in this network and follows the same behavior as with the annotation: it's enabled when set to true and disabled when false or omited in the manifest. It maps to the `k8s.ovn.org/bfd-enabled` annotation.
+  * `podSelector` implements a [set-based](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#set-based-requirement) label selector to filter the pods in the namespace that match this network configuration.
+
+The `status` structure in the CRD defines the following field:
+* `externalGatewayPodIPs` contains a slice of Ipv4 or IPv6 of the pods that are acting as external gateways. This field is currently stored as the annotation `k8s.ovn.org/external-gw-pod-ips` in the namespace and is updated by the controller when there are changes to the gateway pod IPs. This field exists in the status because the dynamic nature of the slice as it is tied to the lifecycle of the pods.
+
+The following yaml examples an instance of the CRD:
+
+```yaml
+kind: MultipleExternalGateway
+apiVersion: k8s.ovn.org/v1
+metadata:
+  name: multiple-external-gateways-configuration
+  namespace: default
+spec:
+  staticGateways: ["9.0.0.1","9.0.0.2"]
+  routingNamespaces: ["default","stage"]
+  podGateways:
+    - network: external-gateway
+      bfdEnabled: true
+      podSelector:
+        - matchLabels:
+          - external-gateway: "true"
+        - matchExpressions:
+          - {key: environment, operator: In, values: [stage]}
+    - network: vpn-network
+      podSelector:
+        - matchLabels:
+          - vpn-egress: "true"
+status:
+  externalGatewayPodIPs: ["10.10.1.2","10.10.1.5"]
+```
+
+### Risks and Mitigations
+N/A
+
+### Drawbacks
+N/A
+
+## Design Details
+
+
+### Test Plan
+
+* Unit test coverage for CR based to at least match use cases in existing annotation based test cases.
```

Comment:
I think we had e2e tests too covering external gateways (if somebody did not kill them :P )

Reply:
:smile: well the idea would be to clone these tests to use the CRs instead. The setup and flows should still apply as long as we are keeping the same feature parity to the annotation implementation.

Reply:
yeah we still have those